### PR TITLE
User module implementations including handlers and authentication logic (Closes #5, Closes #9) 

### LINKED ISSUE #5: Authentication
Implement a user authentication mechanism using a JWT (JSON Web Token) standard.

### LINKED ISSUE #9: Authentication-related endpoints

### PR DESCRIPTION
**Pull request by RedBullDogE**

### CODE REVIEW

File: src/main/java/shortener/httphandler/UserController.java [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+      return HttpResponse.badRequest("Password must contain at least one uppercase character.");
+    }
+
+    if (!userPassword.matches("(.*[a-z].*)")) {
+      return HttpResponse.badRequest("Password must contain at least one lowercase character.");
+    }
+
+    if (!userPassword.matches("(.*[0-9].*)")) {
+      return HttpResponse.badRequest("Password must contain at least one number.");
+    }
```
[the remark this answers is not in the document]
Comment (yugisu) [reply src/main/java/shortener/httphandler/UserController.java:75]: @RedBullDogE Yeah, we could try to move these validation rules into a separate module which throws a ValidationError with related message (or uses other mechanism similar to it)

File: src/main/java/shortener/httphandler/ApiController.java [hunk 2]
```diff
@@ -12,7 +12,7 @@
 /**
  * REST API controller that provides logic for Micronaut framework.
  */
-@Secured(SecurityRule.IS_ANONYMOUS) // TODO: should be setup in the future
+@Secured(SecurityRule.IS_AUTHENTICATED) // TODO: should be setup in the future
```
[the remark this answers is not in the document]
Comment (RedBullDogE) [reply src/main/java/shortener/httphandler/ApiController.java:15]: yeah, sure

File: src/main/java/shortener/httphandler/UserController.java [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+@Controller("/users")
+public class UserController {
+
+  @Inject
+  UserRepository userRepository;
+  @Inject
+  UserSessionRepository userSessionRepository;
+
+
+  record UserData(@JsonProperty("email") String email,
```
[the remark this answers is not in the document]
Comment (RedBullDogE) [reply src/main/java/shortener/httphandler/UserController.java]: @yugisu I tried but it didn't work for some reason

**Comment by yugisu** [comment]
@RedBullDogE Could you resolve conflicts please so we would be able to merge the branch?

File: src/main/java/shortener/database/UserSession.java [hunk 4]
```diff
@@ -6,5 +6,5 @@
  * UserSession record type.
  */
 @JsonAutoDetect(fieldVisibility = JsonAutoDetect.Visibility.ANY)
-public record UsersSession(Integer userId, String token) {
+public record UserSession(Integer userId, String token) {
```
Comment (IINamelessII) [inline src/main/java/shortener/database/UserSession.java:9]: Why? The name was correct...

File: src/main/java/shortener/httphandler/UserController.java [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+  UserRepository userRepository;
+  @Inject
+  UserSessionRepository userSessionRepository;
+
+  @JsonAutoDetect(fieldVisibility = JsonAutoDetect.Visibility.ANY)
+  record UserData(@JsonProperty("email") String email, @JsonProperty("password") String password) {
+
+    public void validate() {
+      if (!email.matches(
+          "^[a-zA-Z0-9_+&*-]+(?:\\.[a-zA-Z0-9_+&*-]+)*@(?:[a-zA-Z0-9-]+\\.)+[a-zA-Z]{2,7}$")) {
```
Comment (IINamelessII) [inline src/main/java/shortener/httphandler/UserController.java:38]: suggest extracting this cumbersome pattern and store it as class'es const, smth like EMAIL_ADDRESS_PATTERN
and use it here (it's good practice for readability of your code)

File: src/main/java/shortener/database/UserSession.java (hunk 4 above)
Comment (IINamelessII) [reply src/main/java/shortener/database/UserSession.java:9]: Nevermind, I've checked all other namings, and it's easier to don't give a fck on it and close the project. But in the future, in your production code - please be mega-careful with names of variables in your code. It really can differ 15-years old projects: a dead one and a maintainable one.

**Review by IINamelessII — DISMISSED** [review]
For the future: Split such PRs into small ones, that can be separated by logic. It's easier to review and to update due to comments :)

**Review by IINamelessII — APPROVED** [review]

**Pull request merged**